File: core/src/main/java/sampler/NativeSampler.java
package sampler;

import java.lang.ref.Cleaner;

public class NativeSampler implements AutoCloseable {
    private final long nativeHandle;
    private final Cleaner.Cleanable cleanable;

    private static final Cleaner cleaner = Cleaner.create();

    private record CleanupAction(long handle) implements Runnable {
        @Override
        public void run() {
            deleteNative(handle);
        }
    }

    public NativeSampler(int sampleRate, int channels) throws Exception {
        long nativeHandle = createNative(sampleRate, channels);
        if (nativeHandle == 0) {
            throw new Exception("Unable to instantiate NativeSampler");
        }
        this.nativeHandle = nativeHandle;
        this.cleanable = cleaner.register(this, new CleanupAction(nativeHandle));
    }

    private static native long createNative(int sampleRate, int channels);

    private static native void setPlaybackSpeedNative(long handle, float factor);

    private static native void setVolumeNative(long handle, float value);

    private static native void startNative(long handle);

    private static native void playNative(long handle, byte[] bytes, int size);

    private static native void stopNative(long handle);

    private static native void deleteNative(long handle);

    public void setPlaybackSpeed(float factor) {
        setPlaybackSpeedNative(nativeHandle, factor);
    }

    public void setVolume(float value) {
        setVolumeNative(nativeHandle, value);
    }

    public void start() {
        startNative(nativeHandle);
    }

    public void play(byte[] data, int size) {
        playNative(nativeHandle, data, size);
    }

    public void stop() {
        stopNative(nativeHandle);
    }

    @Override
    public void close() {
        cleanable.clean();
    }
}

File: core/src/main/kotlin/com/github/numq/klarity/core/sampler/NativeSampler.kt
package com.github.numq.klarity.core.sampler

import com.github.numq.klarity.core.cleaner.NativeCleaner
import java.io.Closeable
import java.util.concurrent.atomic.AtomicLong

internal class NativeSampler(sampleRate: Int, channels: Int) : Closeable {
    private object Native {
        @JvmStatic
        external fun create(sampleRate: Int, channels: Int): Long

        @JvmStatic
        external fun setPlaybackSpeed(handle: Long, factor: Float)

        @JvmStatic
        external fun setVolume(handle: Long, value: Float)

        @JvmStatic
        external fun start(handle: Long): Long

        @JvmStatic
        external fun play(handle: Long, buffer: Long, size: Int)

        @JvmStatic
        external fun pause(handle: Long)

        @JvmStatic
        external fun stop(handle: Long)

        @JvmStatic
        external fun delete(handle: Long)
    }

    private val lock = Any()

    private val nativeHandle = AtomicLong(-1L)

    private fun ensureOpen() {
        check(nativeHandle.get() != -1L) { "Native sampler is closed" }
    }

    init {
        synchronized(lock) {
            require(sampleRate > 0) { "Invalid sample rate" }

            require(channels > 0) { "Invalid channels" }

            nativeHandle.set(Native.create(sampleRate = sampleRate, channels = channels))

            require(nativeHandle.get() != -1L) { "Could not instantiate native sampler" }
        }
    }

    private val cleanable = NativeCleaner.cleaner.register(this) {
        synchronized(lock) {
            ensureOpen()

            Native.delete(handle = nativeHandle.get())

            nativeHandle.set(-1L)
        }
    }

    fun setPlaybackSpeed(factor: Float) = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.setPlaybackSpeed(handle = nativeHandle.get(), factor = factor)
        }
    }

    fun setVolume(value: Float) = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.setVolume(handle = nativeHandle.get(), value = value)
        }
    }

    fun start() = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.start(handle = nativeHandle.get())
        }
    }

    fun play(buffer: Long, size: Int) = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.play(handle = nativeHandle.get(), buffer = buffer, size = size)
        }
    }

    fun pause() = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.pause(handle = nativeHandle.get())
        }
    }

    fun stop() = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.stop(handle = nativeHandle.get())
        }
    }

    override fun close() = cleanable.clean()
}